shell/client/grainview.js

```diff
@@ -748,3 +756,112 @@ const onceConditionIsTrue = (condition, continuation) => {
     });
   });
 };
+
+window.addEventListener("unload", () => {
+  // If more than one grain is open, save a list to restore later. We don't save if just one grain
+  // because people who like to open grains in separate browser tabs probably don't want this
+  // feature. Also, anonymous users who can only open one grain at a time (because they have no
+  // sidebar) would probably be surprised to find the grain re-open if they return to Sandstorm
+  // later.
+
+  // Don't save anything if the user isn't logged in, because users who aren't logged in can't
+  // see the sidebar and probably would be surprised that Sandstorm remembers what they had
+  // opened.
+  if (!Meteor.userId()) return;
+
+  const grains = globalGrains.get();
+  const key = "openGrains-" + SHA256(window.location);
+
+  const old = Meteor._localStorage.getItem(key);
+  if (old) {
+    const oldParsed = JSON.parse(old);
+
+    if (oldParsed.time > Date.now() - 5000) {
+      // Crap. It seems that some other tab was closed in the last 5 seconds that had the same
+      // URL (perhaps a common one like "/apps"). We have no way to distinguish our tab from this
+      // other tab. Rather than arbitrarily clobber one tab's grains list with the other -- which
+      // will likely confuse the user, opening grains in multiple places that weren't previously --
+      // we will have to give up and not restore anything. :(
+      Meteor._localStorage.setItem(JSON.stringify({time: Date.now(), grains: []}));
```

Comment:
This is missing `key`: should be `Meteor._localStorage.setItem(JSON.stringify(key, {time: Date.now(), grains: []}));`